generator/src/main/java/com/algorand/sdkutils/listeners/JavaGenerator.java

```diff
@@ -1,4 +1,933 @@
+            javaQueryWriter.addQueryProperty(type, false, true, false);
+            break;
+        case BODY_CONTENT:
+            javaQueryWriter.addQueryProperty(type, false, false, true);
+            break;
+        default:
+            throw new RuntimeException("Unimplemented event for TypeDef! " + event);
+        }
+    }
+
+    @Override
+    public void onEvent(Events event, StructDef sDef) {
+        switch(event) {
+        case NEW_MODEL:      
+            javaModelWriter.newModel(sDef, this.modelPackage);
+            break;
+        case NEW_RETURN_TYPE:
+            javaModelWriter.newModel(sDef, this.modelPackage);
+            break;
+        default:
+            throw new RuntimeException("Unemplemented event for StructDef! " + event);
+        }
+
+    }
+
+    public void terminate() {
+        javaModelWriter.close();
+        javaModelWriter = null;     
+
+        generateClientFile(
+                clientName,
+                generatedPathsImports,
+                generatedPathsEntries,
+                commonPackage,
+                commonPath,
+                tokenName,
+                tokenOptional);
+
+        generateEnumClasses(this.modelPath, this.modelPackage);
+
+    }
+
+
+    /**
+     * Generate the client which wraps up all the builders, accepts the host/port/token, etc.
+     *
+     * clientName    - IndexerClient
+     * importLines   - The contents of xxxxxImports.txt
+     * paths         - The file at genRoot + xxxxxPaths.txt
+     * packageName   - "com.algorand.algosdk.v2.client.common"
+     * packagePath   - "src/main/java/com/algorand/algosdk/v2/client/common"
+     * tokenName     - "X-Indexer-API-Token"
+     * tokenOptional - Indicates that the token is optional and two constructors should be created.
+     *
+     * @param clientName Name of the client class. i.e. IndexerClient
+     * @param importLines Lines to be added to the import section of the template. Generated from a previous step.
+     * @param paths The generated methods for accessing the paths. Generated from a previous step.
+     * @param packageName Name of the package containing the client.
+     * @param packagePath Path where the client will go.
+     * @param tokenName Name of the token used for this application. i.e. X-Algo-API-Token
+     * @param tokenOptional Whether or not a no-token version of the constructor should be created.
+     */
+    private static void generateClientFile(
+            String clientName, 
+            StringBuilder importLines, 
+            StringBuilder paths, 
+            String packageName, 
+            String packagePath, 
+            String tokenName, 
+            Boolean tokenOptional) {
+
+        if (packagePath.endsWith("/")) {
+            new Exception("Path shouldn't have a trailing slash.").printStackTrace();;
+        }
+
+        importLines.append("import com.algorand.algosdk.crypto.Address;");
+
+        StringBuffer sb = new StringBuffer();
```

Comment:
Use `StringBuilder`, no need for thread safety here.

Reply:
fixed.